_studio/shared/umc/codec/mpeg2_dec/src/umc_mpeg2_splitter.cpp

```diff
@@ -0,0 +1,240 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "umc_defs.h"
+
+#if defined (MFX_ENABLE_MPEG2_VIDEO_DECODE)
+
+#include "umc_media_data.h"
+#include "umc_mpeg2_splitter.h"
+
+namespace UMC_MPEG2_DECODER
+{
+
+    void RawHeaderIterator::LoadData(UMC::MediaData* source)
+    {
+        m_source = source;
+        m_nalu = RawUnit(); // reset
+
+        if (!m_source)
+        {
+            m_nalu = OnEOS();
+            return;
+        }
+
+        m_nalu = FindUnit(m_source);
+
+        if ( (-1 == m_nalu.type) &&  // didn't find complete unit
+                m_cache.size() &&       // but have beginning of the unit
+                !((m_source->GetFlags() & UMC::MediaData::FLAG_VIDEO_DATA_NOT_FULL_FRAME)) ) // and complete frame flag is set
+        {
+            m_nalu = OnEOS();
+        }
+    }
+
+    RawHeaderIterator & RawHeaderIterator::operator++()
+    {
+        m_nalu = RawUnit(); // reset
+
+        if (!m_source)
+        {
+            m_nalu = OnEOS();
+            return *this;
+        }
+
+        m_nalu = FindUnit(m_source);
+
+        if ( (-1 == m_nalu.type) &&  // didn't find complete unit
+                m_cache.size() &&       // but have beginning of the unit
+                !((m_source->GetFlags() & UMC::MediaData::FLAG_VIDEO_DATA_NOT_FULL_FRAME)) ) // and complete frame flag is set
+        {
+            m_nalu = OnEOS();
+        }
+
+        return *this;
+    }
+
+    RawUnit RawHeaderIterator::FindUnit(UMC::MediaData* in)
+    {
+        uint8_t* begin = (uint8_t*) in->GetDataPointer();
+        uint8_t* end   = begin + in->GetDataSize();
+        uint32_t readSize = 0;
+        RawUnit unit;
+
+        if (m_needCleanCache)
+        {
+            m_cache.clear();
+            m_needCleanCache = false;
+            m_pts = -1;
+        }
+
+        // We have cached and got new piece of data
+        if (m_cache.size())
+        {
+            uint8_t* unitStart = FindStartCode(begin, end); // Check if new data have any start codes
+            if (unitStart) // Start code was found. Need to close the cached unit
+            {
+                readSize = unitStart - begin; // Up to found startcode
+                m_cache.insert(m_cache.end(), (uint8_t *)begin, (uint8_t *)begin + readSize); // load data to the cache
+
```

Comment:
Probably with m_cahce.insert(m_cache.end(), 3, 0); // append 3 zero bytes
can avoid checking end of data with every GetBits.
It would also require similar appending in FindRawUnit.

Reply:
Agree, but let's leave this optimization later because it requires discussion. E.g. my proposal is to remove checks in GetBits and make one Check call after parsing the whole unit.